File: resources/recipes/data/flatten-unflatten-objects/flatten.py
"""Flatten and unflatten nested objects in Python.

Usage:
    python flatten.py
"""

import json
import re
from typing import Any


def flatten(obj: Any, separator: str = ".", prefix: str = "") -> dict:
    result = {}
    if isinstance(obj, dict):
        for key, value in obj.items():
            new_key = f"{prefix}{separator}{key}" if prefix else key
            result.update(flatten(value, separator, new_key))
    elif isinstance(obj, list):
        for index, value in enumerate(obj):
            new_key = f"{prefix}[{index}]"
            result.update(flatten(value, separator, new_key))
    else:
        result[prefix] = obj
    return result


def _set(node, parts, value):
    for i, part in enumerate(parts[:-1]):
        next_is_index = parts[i + 1].isdigit()
        if isinstance(node, list):
            index = int(part)
            while len(node) <= index:
                node.append(None)
            if node[index] is None:
                node[index] = [] if next_is_index else {}
            node = node[index]
        else:
            if part not in node:
                node[part] = [] if next_is_index else {}
            node = node[part]

    last = parts[-1]
    if isinstance(node, list):
        index = int(last)
        while len(node) <= index:
            node.append(None)
        node[index] = value
    else:
        node[last] = value


def unflatten(flat: dict, separator: str = ".") -> Any:
    result = {}
    split_re = re.compile(re.escape(separator) + r"|\[|\]")
    for key, value in flat.items():
        parts = [p for p in split_re.split(key) if p]
        _set(result, parts, value)
    return result


if __name__ == "__main__":
    with open("sample.json") as f:
        nested = json.load(f)

    flat = flatten(nested)
    print("=== Flattened ===")
    for key, value in flat.items():
        print(f"  {key}: {value}")

    restored = unflatten(flat)
    print("\n=== Unflattened ===")
    print(json.dumps(restored, indent=2))

    # Verify round-trip
    flat_again = flatten(restored)
    assert flat == flat_again, "Round-trip failed!"
    print("\n✓ Round-trip verified: flatten(unflatten(flat)) == flat")

File: resources/recipes/data/flatten-unflatten-objects/flatten.js
// Flatten and unflatten nested objects in JavaScript.
// Usage: node flatten.js

const fs = require("fs");

function flatten(obj, separator = ".", prefix = "") {
  const result = {};

  if (obj !== null && typeof obj === "object" && !Array.isArray(obj)) {
    for (const [key, value] of Object.entries(obj)) {
      const newKey = prefix ? `${prefix}${separator}${key}` : key;
      Object.assign(result, flatten(value, separator, newKey));
    }
  } else if (Array.isArray(obj)) {
    obj.forEach((value, index) => {
      const newKey = `${prefix}[${index}]`;
      Object.assign(result, flatten(value, separator, newKey));
    });
  } else {
    result[prefix] = obj;
  }

  return result;
}

function _set(node, parts, value) {
  for (let i = 0; i < parts.length - 1; i++) {
    const part = parts[i];
    const nextIsIndex = /^\d+$/.test(parts[i + 1]);
    if (Array.isArray(node)) {
      const index = Number(part);
      while (node.length <= index) node.push(null);
      if (node[index] === null) {
        node[index] = nextIsIndex ? [] : {};
      }
      node = node[index];
    } else {
      if (!(part in node)) {
        node[part] = nextIsIndex ? [] : {};
      }
      node = node[part];
    }
  }

  const last = parts[parts.length - 1];
  if (Array.isArray(node)) {
    const index = Number(last);
    while (node.length <= index) node.push(null);
    node[index] = value;
  } else {
    node[last] = value;
  }
}

function unflatten(flat, separator = ".") {
  const result = {};
  const esc = separator.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
  const splitRe = new RegExp(`${esc}|[\\[\\]]`, "g");

  for (const [key, value] of Object.entries(flat)) {
    const parts = key.split(splitRe).filter(Boolean);
    _set(result, parts, value);
  }

  return result;
}

// Run
const nested = JSON.parse(fs.readFileSync("sample.json", "utf8"));

const flat = flatten(nested);
console.log("=== Flattened ===");
for (const [key, value] of Object.entries(flat)) {
  console.log(`  ${key}: ${value}`);
}

const restored = unflatten(flat);
console.log("\n=== Unflattened ===");
console.log(JSON.stringify(restored, null, 2));

// Verify round-trip
const flatAgain = flatten(restored);
const flatStr = JSON.stringify(flat);
const flatAgainStr = JSON.stringify(flatAgain);
if (flatStr !== flatAgainStr) {
  console.error("✗ Round-trip failed!");
  process.exit(1);
}
console.log("\n✓ Round-trip verified: flatten(unflatten(flat)) == flat");
